store_mux.go

```diff
@@ -0,0 +1,254 @@
+	name       string
+	wildcard   map[string]Store
+	registry   map[string]Store
+	repository map[string]Store
+	fallback   Store
+}
+
+// NewStoreMux creates a new [StoreMux].
+func NewStoreMux(name string) *StoreMux {
+	return &StoreMux{
+		name: name,
+	}
+}
+
+// Name is the name of the store.
+func (s *StoreMux) Name() string {
+	return s.name
+}
+
+// Resolve resolves to a descriptor for the given artifact reference.
+func (s *StoreMux) Resolve(ctx context.Context, ref string) (ocispec.Descriptor, error) {
+	store, err := s.storeFromReference(ref)
+	if err != nil {
+		return ocispec.Descriptor{}, err
+	}
+	return store.Resolve(ctx, ref)
+}
+
+// ListReferrers returns the immediate set of supply chain artifacts for the
+// given subject reference.
+// Note: This API supports pagination. fn should be set to handle the
+// underlying pagination.
+func (s *StoreMux) ListReferrers(ctx context.Context, ref string, artifactTypes []string, fn func(referrers []ocispec.Descriptor) error) error {
+	store, err := s.storeFromReference(ref)
+	if err != nil {
+		return err
+	}
+	return store.ListReferrers(ctx, ref, artifactTypes, fn)
+}
+
+// FetchBlob returns the blob by the given reference.
+// WARNING: This API is intended to use for small objects like signatures,
+// SBoMs.
+func (s *StoreMux) FetchBlob(ctx context.Context, repo string, desc ocispec.Descriptor) ([]byte, error) {
+	store, err := s.storeFromRepository(repo)
+	if err != nil {
+		return nil, err
+	}
+	return store.FetchBlob(ctx, repo, desc)
+}
+
+// FetchManifest returns the referenced manifest as given by the descriptor.
+func (s *StoreMux) FetchManifest(ctx context.Context, repo string, desc ocispec.Descriptor) ([]byte, error) {
+	store, err := s.storeFromRepository(repo)
+	if err != nil {
+		return nil, err
+	}
+	return store.FetchManifest(ctx, repo, desc)
+}
+
+// Register registers a store for the given pattern.
+func (s *StoreMux) Register(pattern string, store Store) error {
+	if pattern == "" {
+		return errors.New("pattern is required")
+	}
+	if store == nil {
+		return errors.New("store is required")
+	}
+
+	// check for FQDN pattern
+	if strings.Contains(pattern, "/") {
+		return s.registerRepository(pattern, store)
+	}
+
+	// check for registry pattern
+	return s.registerRegistry(pattern, store)
+}
+
+// registerRegistry registers a store for the given registry pattern.
+func (s *StoreMux) registerRegistry(pattern string, store Store) error {
+	ref := registry.Reference{
+		Registry: pattern,
+	}
+	if err := ref.ValidateRegistry(); err != nil {
+		return fmt.Errorf("invalid pattern %q: %v", pattern, err)
+	}
+
+	switch strings.Count(pattern, "*") {
+	case 0:
+		if s.registry == nil {
+			s.registry = map[string]Store{
+				pattern: store,
+			}
+		} else {
+			s.registry[pattern] = store
+		}
+	case 1:
+		if !strings.HasPrefix(pattern, "*.") {
+			return fmt.Errorf("invalid pattern %q: wildcard DNS records must start with '*.'", pattern)
+		}
+		pattern = pattern[2:]
```

Comment:
one edge case for cutting it off is that if users provide `*.example.com` and `example.com`, one would override the other.